### PR TITLE
feat!: `flavor` iteration, rework `--overrides` & add examples

### PR DESCRIPTION
**Pull request by sgoudham**
Still some outstanding questions / design decisions to hash out.

I've added a new `nested.hbs` file that does currently render, but I'm not quite sure how frontmatter should be handled when passing in the case of the `all` argument. I don't _think_ we want to treat the entire file as being under the `{{#each}}` nested object but maybe we do?

BEGIN_COMMIT_OVERRIDE
feat: support single files with all flavors
feat!: add new `--overrides` flag
tests: add integration tests
tests: add unit tests for parsing frontmatter
docs(README): refactor into table format
docs(README): add docs on single file generation
END_COMMIT_OVERRIDE

### CODE REVIEW

**Comment by sgoudham** [comment]
From further discussion on discord with @nullishamy, I've also found that using `HashMap` here doesn't guarantee any order (which makes sense) 

This isn't great for us because I'd imagine we always, if not most of the time, want to control the ordering of how the flavours and colours are listed. With the hashmap implementation, I'm always getting Frappe at the start instead of Latte on READMEs.

Potential solutions might be to use Vecs and/or ordered Hashmaps?

**Comment by nekowinston** [comment]
Yeah, I noticed that the order wasn't guaranteed, #104 was pretty rushed to just add the basic functionality without spending much time doing Rust.
IIRC I used `BTreeMap` from stdlib for ordered stuff before, but I'll trust your judgement with the Rust implementation.

As for the frontmatter, it would probably  make sense if we enforced explicit color overrides, like in VSCode/nvim etc, e.g.
```yaml
colorOverrides:
  mocha: {}
  macchiato: {}
  frappe: {}
  latte: {}
  all: {}
```
The other frontmatter can probably be safely assumed to be global, since `isLight`/`isDark` are utility functions, rather than data.

**Comment by sgoudham** [comment]
@nekowinston I did have a look at `BTreeMap` but I didn't find anything that would allow me to pass in a custom sorter for the `colors`, and therefore opted for the `IndexMap`. Let me know if I missed a function like that though.

When you say explicit color overrides, do you mean removing them as a command line flag and making them only available in the frontmatter?

Also what I meant in my initial comment is that I'd still like to _alias_ colours to variable names so that it can act like documentation. 

E.g. With the new changes, I can configure the Slack port as follows:

```
---
weird: "F8F8FA"
---
...
### Theme Strings

{{#each flavors}}
- **{{titlecase flavor}}**{{#if @first}} (Make sure to enable light mode!){{/if}}
  \`\`\`
  #{{uppercase base}},#{{../weird}},#{{uppercase mauve}},#{{uppercase base}},#{{uppercase crust}},#{{uppercase text}},#{{uppercase mauve}},#{{uppercase maroon}},#{{uppercase base}},#{{uppercase text}}
  \`\`\`
{{/each}}
```

But I'd prefer an approach similar to how the slack port currently looks where the variable names are mapped to the colours in the frontmatter:

```
---
column_bg: "{{base}}"
weird: "F8F8FA"
active_item: "{{mauve}}"
active_item_text: "{{base}}"
hover_item: "{{crust}}"
text_color: "{{text}}"
active_presence: "{{mauve}}"
mention_badge: "{{maroon}}"
top_nav_bg: "{{base}}"
top_nav_text: "{{text}}"
---
...
{{#each flavors}}
- **{{titlecase flavor}}**{{#if @first}} (Make sure to enable light mode!){{/if}}
  \`\`\`
  #{{uppercase column_bg}},#{{../weird}},#{{uppercase active_item}},#{{uppercase active_item_text}},#{{uppercase hover_item}},#{{uppercase text_color}},#{{uppercase active_presence}},#{{uppercase mention_badge}},#{{uppercase top_nav_bg}},#{{uppercase top_nav_text}}
  \`\`\`
{{/each}}
```

The approach I've defined is impossible, at the minute, as the frontmatter is rendered before the template itself. Of course, when you set it the `FLAVOR` to `all`, there is no global `{{base}}`. 

I suppose what I'd be interested is maybe a helper of some sort to allow this mapping/aliasing of a variable name to a colour?

**Comment by nekowinston** [comment]
> @nekowinston I did have a look at `BTreeMap` but I didn't find anything that would allow me to pass in a custom sorter for the `colors`, and therefore opted for the `IndexMap`. Let me know if I missed a function like that though.

I wrote that comment before I saw your indexmap commit, consider that resolved 🙃 

> 
> When you say explicit color overrides, do you mean removing them as a command line flag and making them only available in the frontmatter?

Yeah, since you can still set them via JSON in `stdin` / JSON file arguments.
I'd like it if we could have a single option that works the same across all templated ports, whether they're using flavor iteration or not; so explicitly having a single colorOverrides object would help with that.

> Also what I meant in my initial comment is that I'd still like to _alias_ colours to variable names so that it can act like documentation.

That makes sense, terminal ports are a good use case here as well; they can share a common frontmatter like
```yaml
bg: {{ base }}
fg: {{ text }}
sel: {{ surface2 }}
color0: {{ darklight subtext0 surface1 }}
color1: {{ red }}
color2: {{ green }}
color3: {{ yellow }}
color4: {{ blue }}
color5: {{ pink }}
color6: {{ sky }}
color7: {{ darklight subtext1 surface2 }}
color8: {{ darklight surface2 subtext0 }}
color9: {{ red }}
color10: {{ green }}
color11: {{ yellow }}
color12: {{ blue }}
color13: {{ pink }}
color14: {{ sky }}
color15: {{ darklight surface1 subtext1 }}
```
That would require dealing with the current flavor context as well, though.

**Comment by sgoudham** [comment]
@nullishamy and I were discussing overrides earlier and we failed to see how the overrides object could be defined in the frontmatter for single flavours. i.e.

Given this command here in [lazygit](https://github.com/catppuccin/lazygit/blob/4ee9533a77d94e8609522fcb6d2beba1e61befbf/justfile#L13-L14):
```justfile
@{{whiskers_cmd}} {{template_path}} {{flavor}} --override accent="{{accent}}" > {{output_base}}/{{flavor}}/{{accent}}.yml
```

The `--override accent=` flag allows for the template to specify a [placeholder variable](https://github.com/catppuccin/lazygit/blob/4ee9533a77d94e8609522fcb6d2beba1e61befbf/lazygit.hbs#L2) that can be overridden as each [accent colour is passed in](https://github.com/catppuccin/lazygit/blob/4ee9533a77d94e8609522fcb6d2beba1e61befbf/justfile#L18). We weren't really sure how this model could work if all the overrides were specified in the template itself?

I did mention that it feels a bit odd to rely on the overrides flag to generate files with different accents. There's something in me that _feels_ odd about it but maybe that's just me.

**Comment by sgoudham** [comment]
The latest commit that I've just pushed ([`9759473` (#111)](https://github.com/catppuccin/toolbox/pull/111/commits/9759473f1503de9f22acb8f88b54ef67b084d3cd)) allows the frontmatter to be parsed for every single flavour, meaning that we can now successfully recreate the markdown table with every flavour and colour in the correct order.

However, there's still some things to hash out / do:

- [ ] How do we want to handle command line overrides for all flavors? 
- [ ] As the entire frontmatter is being parsed under each the context of each flavour, no frontmatter can be defined in the root context. We need to be able to detect if the frontmatter variable doesn't include `{{ }}` - if it doesn't then we can avoid duplicating it across each flavour.
- [ ] Write tests

**Comment by nekowinston** [comment]
> I've added support for doing command line overrides into the root level context but really don't want to bother with parsing command line overrides when it's multiple flavors, that's where the yaml approach would come in handy. Maybe we should just disallow command line overrides when flavor is set to all?

Yeah

> would probably make sense if we enforced explicit color overrides, like in VSCode/nvim 

in vscode, the overrides are applied like this:
https://github.com/catppuccin/vscode/blob/3cb3529c5c22ea47203f5e0e6e357b6b2623cfce/src/theme/index.ts#L39-L43

Taking the library palette, combining it with `all:` first, then the flavor. So
```json
{
  "colorOverrides": {
     "all": {
       "base": "#000000"
     },
     "latte": {
       "base": "#ffffff"
     }
   }
 }
 ```
 
 would result in Mocha, Macchiato, Frappe having a black background while Latte keeps white.

`whiskers --override '{"colorOverrides":{"all":{"base":"#000000"},"latte":{"base":"#ffffff"}}}'`

**Comment by sgoudham** [comment]
I'm still unsure on how overrides for single flavours could be defined in the yaml as specified in https://github.com/catppuccin/toolbox/pull/111#issuecomment-1859293310 @nekowinston 

The lazygit port essentially relies on a placeholder variable to be overridden on every single invocation for all the accents, having it defined in the yaml doesn't make sense for how that is currently implemented.

**Comment by sgoudham** [comment]
Iterating over accents like that _almost feels_ like it should be handled by more of a build system/task runner (which it technically is with the `justfile`) but it also feels weird that the way to generate multiple accent files with whiskers right now is basically with placeholder variables.

**Comment by nekowinston** [comment]
`whiskers mocha` means that Whiskers is already in the `mocha` flavor context.
```yaml
---
accent: "{{ blue }}"
---
theme:
  activeBorderColor:
    - '#{{ accent }}'
```
"{{ blue }}" is defining the default here; `whiskers mocha --accent "{{ pink }}"` would use `mocha.pink`.

**Comment by nekowinston** [comment]
> Iterating over accents like that _almost feels_ like it should be handled by more of a build system/task runner (which it technically is with the `justfile`) but it also feels weird that the way to generate multiple accent files with whiskers right now is basically with placeholder variables.

I do agree tho that I'd prefer if we didn't have to write absolutely cursed build files, this justfile seems like a step down from saner build scripts. I'd like it if whiskers could handle the "build" similar to how other build systems deal with `<cpu_architecture>-<operating-system>` matrices.

**Comment by sgoudham** [comment]
Agreed, although I think we're doing a good job of trying to integrate it in a variety of different settings and seeing the pain points so we can iteratively improve the design. #112 should really help with this.

**Comment by sgoudham** [comment]
I think you can actually omit `all` in the overrides yaml because the `{{}}` variables act like the `all` context anyways.

I'm going to try and get the following syntax parsed for the single file support:

```yaml
---
# Set default accent color
accent: "{{mauve}}"
# Set default variables in the root context
user: "@sgoudham"
overrides:
  # And you can override variables for specific flavors
  frappe:
    name: "Frappé"
    accent: "{{blue}}"
  mocha:
    accent: "{{green}}"
    user: "@nekowinston"
---
```

Unfortunately it's going to probably going to take some trial and error for me to get right lol

**Comment by nekowinston** [comment]
No my point is that `--overrides` just get merged with the frontmatter as-is, which reduces complexity.

**Comment by sgoudham** [comment]
I suppose the con is that you might potentially have to write out a long JSON string when calling whiskers, but somewhat mitigated by common tools like `jq` / `yq` ?

**Comment by sgoudham** [comment]
I suppose I'm just confused at a potential situation in the frontmatter like below:

```yaml
---
# Set default accent color
accent: "{{mauve}}"
# Set default variables in the root context
user: "@sgoudham"
overrides:
  all:
    accent: "?"
  # And you can override variables for specific flavors
  frappe:
    name: "Frappé"
    accent: "{{blue}}"
  mocha:
    accent: "{{green}}"
    user: "@nekowinston"
---
``` 

You wouldn't define the `all` override if you're already in the single file mode... right?

**Comment by nekowinston** [comment]
> I suppose the con is that you might potentially have to write out a long JSON string when calling whiskers
At that point you're probably customizing a theme rather than authoring it, I'd hope; otherwise our build tool is pretty bad 🙃 
The upside here is that it's at least consistent across ports, I'm mainly thinking of Nix here, e.g. `whiskers --override "${builtins.toJSON myoptions}"`.

That's why I said 
> I'd like it if whiskers could handle the "build" similar to how other build systems deal with <cpu_architecture>-<operating-system> matrices.

I'd love if we could also get generating multiple accents working later, with standardized, generated filenames.

**Comment by sgoudham** [comment]
> I'm mainly thinking of Nix here,

You? The most nix-pilled user on GitHub? Thinking of nix? I am shaken to my very core

<details>
<summary>Context</summary>

image
</details>

Yeah I think a lot of my confusion came from the fact that we're wanting to keep the frontmatter the same even though the overrides section would only be valid in single file mode vs the other modes. I'll have a go at reworking all the overrides stuff to take in a JSON string and merging that with the existing frontmatter since it's only taking in key/value pairs atm.

File: whiskers/src/frontmatter.rs [hunk 1]
```diff
@@ -16,37 +21,153 @@ fn split(template: &str) -> Option<(&str, &str)> {
...
+            .into_iter()
+            .filter(|(k, _)| COLOR_NAMES.iter().any(|s| s == k))
+            .collect::<Map>();
+        if !colours.is_empty() {
+            merged_mut.insert("colors".to_string(), Value::from(colours));
+        }
+
+        // Also, this code isn't performing any validation to check if the override
+        // variables exist beforehand, suppose we need to decide if that's a feature
+        // or bug?
```
Comment (sgoudham) [inline whiskers/src/frontmatter.rs:78]: I actually like not necessarily validating everything since it makes the model more flexible, maybe the only thing to validate is that the `overrides` object always only has either `latte`, `frappe`, `macchiato` or `mocha` at the first level.

File: whiskers/src/main.rs [hunk 2]
```diff
@@ -69,63 +78,75 @@ struct Args {
...
+    // It's quite annoying to differentiate as it would involve using something like
+    // regex to detect if a variable contains "{{ }}" and hold that in a separate
+    // object to extend the "merged" map with here, but it would mean that we
+    // could define variables that can be used outside the "each" loop.
+
+    // if let Some(root) = root_frontmatter {
+    //     let obj = merged.as_object_mut().expect("flavors is an object");
+    //     obj.extend(root);
+    //     obj.extend(overrides_to_map(overrides.into()));
+    // }
```
Comment (sgoudham) [inline whiskers/src/main.rs:139]: I think we do want to allow this, even with the additional complexity it introduces into the codebase. I'm curious what you guys think though @nekowinston @backwardspy (and @nullishamy since we've talked whiskers before)

File: whiskers/tests/cli.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+fn invalid_flavor() -> Result<(), Box<dyn std::error::Error>> {
+    let mut cmd = Command::cargo_bin("whiskers")?;
+    cmd.arg("examples/example/input.hbs").arg("invalid");
+    cmd.assert().failure().stderr(predicate::str::contains(
+        "error: invalid value 'invalid' for '[FLAVOR]'",
+    ));
+    Ok(())
+}
+
+// wanted to use the `test-case` crate to parameterize the test, but it's nightly only
```
Comment (sgoudham) [inline whiskers/tests/cli.rs:26]: It'd be nice if we could switch to nightly but I had a look at the github workflows and I'm not so sure if it's straightforward or not :sweat_smile:

**Comment by sgoudham** [comment]
I think these changes that I'll be introducing will result in a major bump since I'm changing override flag and all the underlying logic. 

In terms of the code, I've got a few questions scattered about the place that I'll revisit. Otherwise, the all flavour mode and single flavor mode do work :tm: 

I've started to include examples of varying complexity to showcase what it can do now. For this PR, I'd like to tidy up the documentation, add more unit tests, add more examples and tidy up the absolute dumpster fire code that I've introduced lol

File: whiskers/tests/cli.rs (hunk 3 above)
Comment (nekowinston) [reply whiskers/tests/cli.rs:37]: We could change it, but is raising the MSRV for tests worth it?
If you really think so, we could maybe feature guard it against a new `nightly` feature and check it on CI, and release on stable; expecting people to pull in rust from nightly for tests feels a bit rough, when even my WASM code works within the limitations of stable.

Comment (sgoudham) [reply whiskers/tests/cli.rs:37]: The test suite I had in mind is similar to what I did for `git-view` -> https://github.com/sgoudham/git-view/blob/2e3ae35a35d03fb70b508fb49b82ee34ab0345f9/src/lib.rs#L785-L787

E.g. here that could look like

```rust
#[test_case("latte"          ; "with flavor latte")]
#[test_case("frappe"         ; "with flavor frappe")]
#[test_case("macchiato"      ; "with flavor macchiato")]
#[test_case("mocha"          ; "with flavor mocha")]
fn some_test(flavor: &str) {
	...
}
```

Just helpful for easily testing all permutations without much duplication/complexity, of course it should still be used sparingly to test the absolute core functionality.

Comment (sgoudham) [reply whiskers/tests/cli.rs]: If it's too much hassle then we don't need to bother at all, I did feel a bit uneasy trying to update things to nightly with wasm stuff kicking around :smile:

Comment (sgoudham) [reply whiskers/tests/cli.rs]: Those "integration" tests are meant to be lightweight anyways, I just wanted to check the binary didn't fall over, the bulk of the logic testing will be at the unit level :+1: 

I'll mark this as resolved.

File: whiskers/src/frontmatter.rs (hunk 1 above)
Comment (sgoudham) [reply whiskers/src/frontmatter.rs]: I actually kinda like this as more of a feature because, for example, what I've been able to do here -> https://github.com/catppuccin/toolbox/blob/feat/flavour-iteration/whiskers/examples/single-file-with-overrides/input.hbs

but I think we might actually need to work the terminology because it's not really doing much "overriding" if it's adding new frontmatter into the template... I don't know though :thinking:

File: whiskers/src/main.rs (hunk 2 above)
Comment (sgoudham) [reply whiskers/src/main.rs]: I could only think of one usecase for this (where you'd maybe want to callout in another part of the document referencing the variable) but honestly I'd like to get this PR wrapped up since it's already so big.

I wouldn't mind revisiting this later on

File: whiskers/src/frontmatter.rs (hunk 1 above)
Comment (sgoudham) [reply whiskers/src/frontmatter.rs]: I've opted to not go for validation because I do like the flexibility of being able to define new variables in the "overrides"

Happy to implement it if we think it's worth it though.

**Comment by sgoudham** [comment]
There's a whole load of stuff still to try and figure out but given how this PR was originally supposed to be a fraction of the size that is now, I think it's a great time to stop :upside_down_face: 

The iteration works and the merge logic works as outlined in the `frontmatter.rs` tests, I'm just going to update the README and hopefully move this PR out of draft. Half of the files are just the generated output files so hopefully this PR isn't the worse to review by just looking at the tests and the examples :+1: 

There are outlining questions / considerations about various edge cases here and there but let's think about them when they start being a pain :D

File: whiskers/tests/cli.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    use assert_cmd::cargo::CommandCargoExt;
+    use predicates::prelude::predicate;
+    use std::fs;
+    use std::process::Command;
+
+    #[test]
+    fn example_file_has_flavor_mocha() -> Result<(), Box<dyn std::error::Error>> {
+        let mut cmd = Command::cargo_bin("whiskers")?;
+        let expected =
+            fs::read_to_string("examples/demo/output/mocha.md").expect("expected file is readable");
```
Comment (nekowinston) [inline whiskers/tests/cli.rs:13]: pretty sure we can use
```rs
let bytes = include_bytes!("examples/demo/output/mocha.md");
```
for all the expected stuff here

**Comment by sgoudham** [comment]
Just have been writing documentation and I've only now just realised that we don't actually need the `all` override at all, because the frontmatter is basically the `all` override already.

Will aim to push everything up soon you can understand it via the README :+1:

**Comment by sgoudham** [comment]
As discussed on discord with @nekowinston, the current overrides are always in the context of the current flavour, and this may not be the exact behaviour that we want. 

However, this PR puts us in a better place than we were before so should be good to get reviewed and merged now :+1: 

Half of the files modified are new examples so it's best to just ignore that directory in the file view and jump to reviewing the `*.rs` files.

**Review by nullishamy — APPROVED** [review]
LGTM. Would have some notes about style/architecture but given that this is still very early stages, I think we can leave it for future refactoring.

**Review by sgoudham — COMMENTED** [review]
This PR has marinated long enough, will merge and we can continue to iterate on it.
